buffer.go

```diff
@@ -0,0 +1,196 @@
+package erasurecode
+
+import (
+	"bytes"
+	"io"
+)
+
+type BufferMatrix struct {
+	b                []byte
+	hdrSize, bufSize int
+	len              int // len of input
+	k                int
+	curBlock         int
+	leftInBlock      int
+	finished         bool
+}
+
+// FragLen returns the size of a "fragment" aligned to a block size (data + header)
+func (b BufferMatrix) FragLen() int {
+	return b.SubGroups() * (b.bufSize + b.hdrSize)
+}
+
+// SubGroups returns the number of blocks inside a single fragment
+func (b BufferMatrix) SubGroups() int {
+	nbBlocks := (b.len + b.bufSize - 1) / b.bufSize
+	nbStripes := (nbBlocks + b.k - 1) / b.k
+	return nbStripes
+}
+
+func (b BufferMatrix) maxLen() int {
+	return (b.SubGroups() * b.k) * (b.bufSize + b.hdrSize)
+}
+
+// NewBufferMatrix returns a new buffer suitable for <len> data and organized
+// such as it can be injected into EncodeMatrixWithBuffer without allocation/copying
+// the data into shards
+func NewBufferMatrix(bufSize int, len int, k int) *BufferMatrix {
+	var b BufferMatrix
+	b.Reset(bufSize, len, k)
+	return &b
+}
+
+// Reset serves the same purpose as NewBufferMatrix but use the existing buffer and
+// tries to avoid allocation of the underlying buffer.
+func (b *BufferMatrix) Reset(bufSize int, len int, k int) {
+	hdrSize := fragmentHeaderSize()
+	b.hdrSize = hdrSize
+	b.bufSize = bufSize
+	b.len = len
+	b.k = k
+	b.leftInBlock = -1
+	b.curBlock = 0
+	b.finished = false
+
+	maxLen := b.maxLen()
+
+	if cap(b.b) <= maxLen {
+		// No internal buffer or buffer is too small, let's replace it
+		b.b, _ = memalign(maxLen, defaultAlign)
+	}
+	b.b = b.b[:]
```

Comment:
What is this syntax  ? you copy yourself from yourself from begin to end ?

Reply:
Good catch, this is actually wrong, this is supposed to be [:0] => aka reset size to 0 without touching capacity.

Reply:
https://play.golang.org/p/4-gNdK4FxVR for a quick test